### ISSUE TITLE
Add hint about --stacktrace to deprecation warnings

### ISSUE THREAD

**Issue by pranasblk** [gradle/gradle#899]
The warning `"Task.leftShift(Closure) method has been deprecated and is scheduled to be removed"` is printed out only if legal syntax of applying plugin `"apply plugin: '...'"` 

The apply plugin: syntax is documented as legal 3.2 
Chapter ["Applying plugins with the buildscript block"](https://docs.gradle.org/current/userguide/userguide_single.html)

### Expected Behavior

No warnings! for 

```
buildscript {
  repositories {
  // ...
 }
}
apply plugin: 'net.researchgate.release'
apply plugin: 'me.champeau.gradle.jmh'

``` expected.

<!--- If you're suggesting a change/improvement, tell us how it should work -->
How it should work: Remove warning for a legal usage of applying plugins.
  
### Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
Expected and current behaviour:
Currently:
1.  prints warning (must not)
2. warning is incorrect the << is not used anywhere in build.gradle

Expected:
No warnings

<!--- If suggesting a change/improvement, explain the difference from current behavior -->
Expected change: Remove the warning

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
Goal: I want to update gradle to 3.2 and explicitly specify binary plugins location from the private (controlled) artifact repository.

<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Context: I want secure environment with fixed performance bugs @ 3.2 version and build without warnings.

### Steps to Reproduce (for bugs)
<!--- Provide a link to a build scan, example project, or an unambiguous -->
<!--- set of steps to reproduce this bug. Include code to reproduce, if relevant -->
Reproduce build with:
```
buildscript {
  repositories {
  // Specify private repo for the binary plugin
      }
    }
  }
  dependencies {
    classpath 'net.researchgate:gradle-release:2.3.5',
              'me.champeau.gradle:jmh-gradle-plugin:0.3.1',
              'net.ltgt.gradle:gradle-apt-plugin:0.8'
  }
}

apply plugin: 'net.researchgate.release'
apply plugin: 'me.champeau.gradle.jmh'
```

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
Gradle 3.2

<!--- A build scan `https://scans.gradle.com/get-started?type=project` is ideal -->

**Comment by delego-wlritchi** [comment]
Also experiencing this issue; I believe this occurs when the plugin itself contains a use of the deprecated syntax. I would expect to see a warning, but pointing to the plugin's source rather than to the `apply plugin` line.

**Comment by huxi** [comment]
Running `gradle --stacktrace` with this `build.gradle`

```groovy
buildscript {
	repositories {
		maven {
			url "https://plugins.gradle.org/m2/"
		}
	}
	dependencies {
		classpath 'net.researchgate:gradle-release:2.3.5'
		classpath 'me.champeau.gradle:jmh-gradle-plugin:0.3.1'
		classpath 'net.ltgt.gradle:gradle-apt-plugin:0.8'
	}
}

apply plugin: 'net.researchgate.release'
apply plugin: 'me.champeau.gradle.jmh'
```

prints this:
```
The Task.leftShift(Closure) method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use Task.doLast(Action) instead.
        at org.gradle.api.internal.AbstractTask.leftShift(AbstractTask.java:505)
        at org.gradle.api.Task$leftShift$0.call(Unknown Source)
        at net.researchgate.release.ReleasePlugin.apply(ReleasePlugin.groovy:61)
        at net.researchgate.release.ReleasePlugin.apply(ReleasePlugin.groovy)
        at org.gradle.api.internal.plugins.ImperativeOnlyPluginApplicator.applyImperative(ImperativeOnlyPluginApplicator.java:35)
        [..]
```

So the culprit is indeed the [net.researchgate.release](https://plugins.gradle.org/plugin/net.researchgate.release) plugin.

I just tried out the latest `2.4.1` version and it still causes the warning. You could reach out to the developer of the plugin or wait for the next release, hoping it will be addressed.

Gradle is working as expected.

**Comment by huxi** [comment]
I created a PR for the fix: https://github.com/researchgate/gradle-release/pull/204

**Comment by bmuschko** [comment]
@pranasblk The warning message can come from any other plugin you apply to your build script. The root cause might not be your own code. Given that the stacktrace indicate the source I think it should provide sufficient information to indicate the issue. Do you agree that we can close the issue?

**Comment by pranasblk** [comment]
@bmuschko, There are no << in my code.
If gradle was tested on
net.researchgate:gradle-release:2.3.5
me.champeau.gradle:jmh-gradle-plugin:0.3.1
net.ltgt.gradle:gradle-apt-plugin:0.8 with researchgate/gradle-release#204 it can be closed. 

Thanks!

**Comment by bmuschko** [comment]
@pranasblk I don't think we want to introduce a way to suppress the warning. That would defeat the purpose. As soon as we release Gradle 5.0 all builds would break that uses the left shift operator in any form. We'll simply remove the operator from the API. Until then there's plenty of time to align build code _and_ plugins.

**Comment by delego-wlritchi** [comment]
It might help to tweak the error message in this case to indicate that the deprecated syntax is being used by a plugin. That way people don't get the idea that the error message is wrong.

Right now, the error message looks like this:
```
The Task.leftShift(Closure) method has been deprecated and is scheduled to be removed in Gradle 5.0. Please use Task.doLast(Action) instead.
	at build_chah3zei0quee2sahreicohc.run(/path/to/my/build.gradle:34)
```
Which really does feel like it's saying that my build script uses this syntax.

**Comment by bmuschko** [comment]
@delego-wlritchi The behavior you are describing applies to all deprecation messages emitted by Gradle. I'd be surprised if you ran into this for the first time. I agree that the error message could be improved by being more specific about the source of the deprecation. However, figuring that out might be a non-trivial change. Would you be interested in working on improved deprecation logging?

**Comment by bmuschko** [comment]
@pranasblk Can we close the issue?

**Comment by pranasblk** [comment]
@bmuschko The warning message can be improved. Would you mind keeping in the back-log instead of closing (without resolution). Other users would be able to see know issue and it's progress.

**Comment by bmuschko** [comment]
@pranasblk I can keep the issue open but we are likely not going to work on this anytime soon. Please vote up the issue if you want it to get a higher priority. A proper solution would take into account any deprecation message and not just this specific one. We'd be happy to review and accept contributions from the community in the meantime.

**Comment by Vampire** [comment]
The situation is even worse if you apply the plugin via plugins DSL or as settings plugin, as then you do not even have any location information without `--stacktrace`.

Also as every deprecation warning is only output once, it shades other occurences in other 3rd party plugins and you also cannot fix your own usages until all 3rd parties fixed theirs.

See examples and some solution / improvement suggestions in my duplicate #1269.

**Comment by huxi** [comment]
For the sake of info consolidation:
My PR #699 is tackling (among other things) the issue that deprecation warnings are only printed once.

I'm not sure if trying to make the selection of the right stacktrace element smarter is the right way to tackle the problem of "false" warnings caused by 3rd-party plugins. I think the behavior is OK for the case were an STE is printed pointing to the place where the plugin is applied. (Disclaimer: I wrote the code responsible for printing the single STE so I'm not entirely objective.)

The other situation were no STE is printed is indeed a problem, though. How about printing some instructions like `Use --stacktrace to see the full stacktrace of this warning.` if no build script location can be pin-pointed instead? Or maybe even print those instructions in case of all deprecation warnings.

Only getting the first occurrence of a deprecation warning is indeed a quite critical problem that already caused some grief for me. I tried to help the people developing the `spring-boot-plugin` getting rid of their warnings before the 1.5 release but ultimately failed because one location slipped through the net. (See https://github.com/spring-projects/spring-boot/issues/8154)
Our internal build is actually triggering this warning so I could have told them in a timely manner. But it was unfortunately shadowed by a previous warning of the same type.

**Comment by Katiyman** [comment]
Can this warning be ignored?

**Comment by Vampire** [comment]
@Katiyman as with any deprecated warnings, you can ignore them as long as you don't upgrade to a version that removes the feature. So you better fix it.

**Comment by alessio-dev2rights** [comment]
@Katiyman how @Vampire already said is better to don't ignore it.... My experience..I had the same issue and inside I had the script to autoIncrementBuildNumber() once gradle ignore that of course this will cause an error where no one was able to specified the build number because was always the same. 
My fix has been swap the "<<" with "doLast{doTheActionYouNeed()}"
Before: task myTask << {autoIncrementBuildNumber()}
After: task myTask {doLast{autoIncrementBuildNumber()}}

**Comment by jjohannes** [comment]
> **Add hint about --stacktrace to deprecation warnings**

This has been addressed with #2778. Further improvements to deprecation warnings should be addressed in separate issues.
